src/main/java/fi/aalto/cs/apluscourses/intellij/actions/SubmitExerciseAction.java

```diff
@@ -0,0 +1,186 @@
+        exercise, authentication, project);
+    if (submissionHistory == null) {
+      return;
+    }
+
+    List<Group> groups = tryGetUserGroups(course, authentication, project);
+    if (groups == null) {
+      return;
+    }
+
+    SubmissionViewModel viewModel = new SubmissionViewModel(
+        exercise, submissionHistory, groups, authentication, project);
+
+    if (!new ModuleSelectionDialog(viewModel).showAndGet()
+        || viewModel.getSelectedModule() == null) {
+      return;
+    }
+
+    List<Path> filePaths = tryGetFilePaths(
+        exercise.getFilenames(), viewModel.getSelectedModule(), project);
+    if (filePaths == null) {
+      return;
+    }
+
+    if (new SubmissionDialog(viewModel).showAndGet()) {
+      // Do actual submission
+    }
+  }
+
+  @Nullable
+  private SubmittableExercise tryGetSubmittableExercise(@NotNull Exercise exercise,
+                                                        @NotNull APlusAuthentication authentication,
+                                                        @Nullable Project project) {
+    try {
+      SubmittableExercise submittableExercise = SubmittableExercise.fromExerciseId(
+          exercise.getId(), authentication);
+      if (submittableExercise.getFilenames().isEmpty()) {
+        notifier.notify(new NotSubmittableNotification(), project);
+        return null;
+      }
+      return submittableExercise;
+    } catch (IOException e) {
+      notifyNetworkError(e, project);
+      return null;
+    }
+  }
+
+  @Nullable
+  private SubmissionHistory tryGetSubmissionHistory(@NotNull Exercise exercise,
+                                                    @NotNull APlusAuthentication authentication,
+                                                    @Nullable Project project) {
+    try {
+      return SubmissionHistory.getSubmissionHistory(exercise.getId(), authentication);
+    } catch (IOException e) {
+      notifyNetworkError(e, project);
+      return null;
+    }
+  }
+
+  @Nullable
+  private List<Group> tryGetUserGroups(@NotNull Course course,
+                                       @NotNull APlusAuthentication authentication,
+                                       @Nullable Project project) {
+    try {
+      return Group.getGroups(course, authentication);
+    } catch (IOException e) {
+      notifyNetworkError(e, project);
+      return null;
+    }
+  }
+
+  @CalledWithReadLock
+  @Nullable
+  private List<Path> tryGetFilePaths(@NotNull List<String> filenames,
+                                     @NotNull Module module,
+                                     @Nullable Project project) {
+    Path modulePath = Paths.get(ModuleUtilCore.getModuleDirPath(module));
+    List<Path> filePaths = new ArrayList<>(filenames.size());
+    for (String filename : filenames) {
+      Path filePath = VfsUtil.findFileInDirectory(modulePath, filename);
+      if (filePath == null) {
+        notifier.notify(new MissingFileNotification(module.getName(), filename), project);
```

Comment:
so there is a notification for each missing file, right?
just thinking, that if there's some fault in loading the modules, will it result in dozens of notifications or... :question:

Reply:
The method returns after showing the first notification so only the first missing file gets a notification